src/types/web-resource.ts

```diff
@@ -0,0 +1,147 @@
+import * as TypeUtils from '../type-utils';
+import type {
+	CastNode,
+	ExtractJSONPathAsTextNode,
+	ReferencedFieldNode,
+} from '@balena/abstract-sql-compiler';
+
+export type WebResource = {
+	filename: string;
+	href: string;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+};
+
+export const types = {
+	postgres: 'JSONB',
+	mysql: 'JSON',
+	websql: 'TEXT',
+	odata: {
+		name: 'Self.WebResource',
+		complexType: `\
+<ComplexType Name="WebResource">
+	<Property Name="filename" Nullable="false" Type="Edm.String"/>\
+	<Property Name="href" Nullable="false" Type="Edm.String"/>\
+	<Property Name="contentType" Nullable="true" Type="Edm.String"/>\
+	<Property Name="contentDisposition" Nullable="true" Type="Edm.String"/>\
+	<Property Name="size" Nullable="true" Type="Edm.Int64"/>\
+</ComplexType>`,
+	},
+};
+
+export const nativeProperties = {
+	has: {
+		Filename: (
+			referencedField: ReferencedFieldNode,
+		): ExtractJSONPathAsTextNode => [
+			'ExtractJSONPathAsText',
+			referencedField,
+			['TextArray', ['EmbeddedText', 'filename']],
+		],
+		HRef: (referencedField: ReferencedFieldNode): ExtractJSONPathAsTextNode => [
+			'ExtractJSONPathAsText',
+			referencedField,
+			['TextArray', ['EmbeddedText', 'href']],
+		],
+		'Content Type': (
+			referencedField: ReferencedFieldNode,
+		): ExtractJSONPathAsTextNode => [
+			'ExtractJSONPathAsText',
+			referencedField,
+			['TextArray', ['EmbeddedText', 'contentType']],
+		],
+		'Content Disposition': (
+			referencedField: ReferencedFieldNode,
+		): ExtractJSONPathAsTextNode => [
+			'ExtractJSONPathAsText',
+			referencedField,
+			['TextArray', ['EmbeddedText', 'contentDisposition']],
+		],
+		Size: (referencedField: ReferencedFieldNode): CastNode => [
+			'Cast',
+			[
+				'ExtractJSONPathAsText',
+				referencedField,
+				['TextArray', ['EmbeddedText', 'size']],
+			],
+			'Integer',
+		],
+	},
+};
+
+/**
+ * Converts the data, which comes from the DB as a string or object depending on the
+ * column type, to a WebResource object
+ *
+ * @param data string|object
+ * @returns a WebResource parsed from the DB
+ */
+export const fetchProcessing = (data: any) => {
+	let refData: WebResource;
+	if (data === null) {
+		return data;
+	}
+	if (typeof data === 'string') {
+		try {
+			refData = JSON.parse(data);
+		} catch (e: any) {
+			throw new Error(`can't parse value of type ${typeof data}`);
```

Comment:
In this case the type is not such a useful bit of info because it's guaranteed to be a string, and the error message would be better off indicating that it's invalid JSON